### PR TITLE
[RemoveFinalFromConstRector]

### PR DESCRIPTION
**Pull request by reypm**
- Added a rule to remove `final` from constants when they belong to a `final` class to avoid issues with other static analysis tools like https://github.com/nunomaduro/phpinsightslike where you start getting "errors" like:

> ::error file=/var/www/ACL/Add.php,line=11::* [Unnecessary final modifier] Unnecessary FINAL modifier in FINAL class

```php
// before
final class A {
    final public const AC = 'const';
}

// after
final class A {
    public const AC = 'const';
}
```

### CODE REVIEW

File: config/set/php81.php [hunk 1]
```diff
@@ -27,4 +28,5 @@
     $services->set(NewInInitializerRector::class);
     $services->set(IntersectionTypesRector::class);
     $services->set(NullToStrictStringFuncCallArgRector::class);
+    $services->set(RemoveFinalFromConstRector::class);
```
Comment (samsonasik) [inline config/set/php81.php]: this should be moved to `config/set/coding-style.php`

File: rules-tests/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector/Fixture/some_class.php.inc [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+?>
+-----
+<?php
+
+namespace Rector\Tests\CodingStyle\Rector\ClassConst\RemoveFinalFromConstRector\Fixture;
+
+final class SomeClass
+{
+    public final const NAME = 'value';
```
Comment (samsonasik) [inline rules-tests/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector/Fixture/some_class.php.inc]:
```suggestion
    public const NAME = 'value';
```

File: rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+use Rector\Privatization\NodeManipulator\VisibilityManipulator;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://php.watch/versions/8.1/final-class-const
+ *
+ * @see \Rector\Tests\CodingStyle\Rector\ClassConst\RemoveFinalFromConstRector\RemoveFinalFromConstRectorTest
+ */
+final class RemoveFinalFromConstRector extends AbstractRector
```
Comment (samsonasik) [inline rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php]: the rule needs to implements `Rector\VersionBonding\Contract\MinPhpVersionInterface` so you can add:

```php
    public function provideMinPhpVersion(): int
    {
        return \Rector\Core\ValueObject\PhpVersionFeature::FINAL_CLASS_CONSTANTS;
    }
```

so it ensure check on php 8.1

File: rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (! $parentClass instanceof Class_) {
+            return null;
+        }
+
+        if ($parentClass->isFinal() && $node->isFinal()) {
+            $this->visibilityManipulator->removeFinal($node);
+        }
+
+        return $node;
```
Comment (samsonasik) [inline rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php]: can return `$node` early after remove `final`, so next will return `null` as not changed.

```suggestion
            $this->visibilityManipulator->removeFinal($node);
            return $node;
        }
        
        return null;    
```

File: rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php (hunk 3 above)
Comment (samsonasik) [reply rules/CodingStyle/Rector/ClassConst/RemoveFinalFromConstRector.php]:
```suggestion
final class RemoveFinalFromConstRector extends AbstractRector implements \Rector\VersionBonding\Contract\MinPhpVersionInterface
```

**Comment by samsonasik** [comment]
@reypm it seems ready?

**Comment by reypm** [comment]
@samsonasik yessir, marking it as Ready for Review for you and the rest of the team 😎

**Review by samsonasik — APPROVED** [review]

**Review by p3dry — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you :clap:

**Pull request merged**